ucsb-cs156-s22/s22-6pm-happycows
# Staff PC:  Pull in code to run async jobs, and address backend testing tech debt

# Overview

In this PR, we accomplish three things:

* We add code for an asynchronous job service, originally developed by @btk5h in the project 
  https://github.com/brownfield-team/kanban-populator.   This async job service is needed for
  implementing the features that increase and decrease cow health, as well as milking the cows,
  and noticing with cow health has degraded to the point of cow death.
* We add a controller and controller test that shows how to use the job service, and allows it to be tested through swagger-ui.
* We address some backend test coverage gaps (i.e. some tech debt from W22).

# Screenshots

<img width="774" alt="image" src="https://user-images.githubusercontent.com/1119017/170566857-b410f188-f454-4a74-98f9-986f02c15e12.png">

# Testing

* To test, visit the `/api/jobs/all` endpoint, and see the list of jobs.  It may initially be empty if no-one has tested yet.
* Then, launch a test job via the `/api/jobs/launch/testjob` endpoint. 

  <img width="784" alt="image" src="https://user-images.githubusercontent.com/1119017/170567277-9af57fea-5947-40bf-9ea9-ee0622dd6448.png">
  
  - Use `fail=false` if you want the job to succeed without an exception. 
  - Use `fail=true` if you want to test what happens when a job fails with an exception
  - Enter a value in milliseconds for the job to sleep (this simulates a job that takes a while.)

* Then, check  `/api/jobs/all` endpoint and you should see, after some time, that the job completes.
